### ISSUE TITLE
WENO5 is much slower on GPUs after Julia 1.6 upgrade

### ISSUE THREAD

**Issue by tomchor** [CliMA/Oceananigans.jl#1764]
I've noticed a few weeks ago that my scripts were much slower after the Julia 1.6 upgrade (which is preventing me from upgrading). I thought it was due to my Julia 1.6 installation but after some tests I now think it's an Oceananigans issue, specifically with the WENO5 scheme.

I ran the MWE below in both Julia 1.5 (with Oceananigans version 0.57.1) and Julia 1.6 (tried several Oceananigans versions but specifically for this example I'm using Oceananigans version 0.58.5) using GPUs and the speed difference is pretty huge. The interesting part is that this difference only happens if I use WENO5 with a GPU. If I use the 2nd order centered scheme there is no significant difference in time (I haven't tried other schemes) and if I run the script on CPUs the time difference also appears to be small.

Here's the script:

```julia
using Oceananigans
using Oceananigans.Units
using CUDA: has_cuda
Nx, Ny, Nz = 128, 1600, 64

if has_cuda()
    arch = GPU()
else
    arch = CPU()
    Nx = Int(Nx/4)
    Ny = Int(Ny/4)
    Nz = Int(Nz/4)
end 

topology = (Periodic, Bounded, Bounded)
grid = RegularRectilinearGrid(size=(Nx, Ny, Nz),
                              x=(0, 200),
                              y=(0, 2000),
                              z=(-100, 0),
                              topology=topology)
println("\n", grid, "\n")


model = IncompressibleModel(architecture = arch,
                grid = grid,
                advection = WENO5(),
                timestepper = :RungeKutta3,
                tracers=nothing,
                buoyancy=nothing,
                closure=nothing,
                )
println("\n", model, "\n")



start_time = 1e-9*time_ns()
using Oceanostics: SingleLineProgressMessenger
simulation = Simulation(model, Δt=10seconds,
                        stop_time=10hours,
                        wall_time_limit=23.5hours,
                        iteration_interval=5,
                        progress=SingleLineProgressMessenger(LES=false, initial_wall_time_seconds=start_time),
                        stop_iteration=Inf,)

println("\n", simulation,"\n",)
@info "---> Starting run!\n"
run!(simulation, pickup=false)
```

The output for Julia 1.5:

```julia
[ Info: ---> Starting run!
[ Info: [000.14%] i:      5,     time: 50.000 seconds,     Δt: 10 seconds,     wall time: 1.447 minutes,     adv CFL: 0.00e+00,     diff CFL: 0.00e+00
[ Info: [000.28%] i:     10,     time: 1.667 minutes,     Δt: 10 seconds,     wall time: 1.612 minutes,     adv CFL: 0.00e+00,     diff CFL: 0.00e+00
[ Info: [000.42%] i:     15,     time: 2.500 minutes,     Δt: 10 seconds,     wall time: 1.751 minutes,     adv CFL: 0.00e+00,     diff CFL: 0.00e+00
[ Info: [000.56%] i:     20,     time: 3.333 minutes,     Δt: 10 seconds,     wall time: 1.890 minutes,     adv CFL: 0.00e+00,     diff CFL: 0.00e+00
[ Info: [000.69%] i:     25,     time: 4.167 minutes,     Δt: 10 seconds,     wall time: 2.028 minutes,     adv CFL: 0.00e+00,     diff CFL: 0.00e+00
[ Info: [000.83%] i:     30,     time:  5 minutes,     Δt: 10 seconds,     wall time: 2.167 minutes,     adv CFL: 0.00e+00,     diff CFL: 0.00e+00
[ Info: [000.97%] i:     35,     time: 5.833 minutes,     Δt: 10 seconds,     wall time: 2.307 minutes,     adv CFL: 0.00e+00,     diff CFL: 0.00e+00
[ Info: [001.11%] i:     40,     time: 6.667 minutes,     Δt: 10 seconds,     wall time: 2.446 minutes,     adv CFL: 0.00e+00,     diff CFL: 0.00e+00
[ Info: [001.25%] i:     45,     time: 7.500 minutes,     Δt: 10 seconds,     wall time: 2.585 minutes,     adv CFL: 0.00e+00,     diff CFL: 0.00e+00
[ Info: [001.39%] i:     50,     time: 8.333 minutes,     Δt: 10 seconds,     wall time: 2.724 minutes,     adv CFL: 0.00e+00,     diff CFL: 0.00e+00
[ Info: [001.53%] i:     55,     time: 9.167 minutes,     Δt: 10 seconds,     wall time: 2.863 minutes,     adv CFL: 0.00e+00,     diff CFL: 0.00e+00
[ Info: [001.67%] i:     60,     time: 10.000 minutes,     Δt: 10 seconds,     wall time: 3.002 minutes,     adv CFL: 0.00e+00,     diff CFL: 0.00e+00
[ Info: [001.81%] i:     65,     time: 10.833 minutes,     Δt: 10 seconds,     wall time: 3.141 minutes,     adv CFL: 0.00e+00,     diff CFL: 0.00e+00
[ Info: [001.94%] i:     70,     time: 11.667 minutes,     Δt: 10 seconds,     wall time: 3.280 minutes,     adv CFL: 0.00e+00,     diff CFL: 0.00e+00
[ Info: [002.08%] i:     75,     time: 12.500 minutes,     Δt: 10 seconds,     wall time: 3.419 minutes,     adv CFL: 0.00e+00,     diff CFL: 0.00e+00
[ Info: [002.22%] i:     80,     time: 13.333 minutes,     Δt: 10 seconds,     wall time: 3.558 minutes,     adv CFL: 0.00e+00,     diff CFL: 0.00e+00
[ Info: [002.36%] i:     85,     time: 14.167 minutes,     Δt: 10 seconds,     wall time: 3.697 minutes,     adv CFL: 0.00e+00,     diff CFL: 0.00e+00
[ Info: [002.50%] i:     90,     time: 15.000 minutes,     Δt: 10 seconds,     wall time: 3.836 minutes,     adv CFL: 0.00e+00,     diff CFL: 0.00e+00
[ Info: [002.64%] i:     95,     time: 15.833 minutes,     Δt: 10 seconds,     wall time: 3.975 minutes,     adv CFL: 0.00e+00,     diff CFL: 0.00e+00
[ Info: [002.78%] i:    100,     time: 16.667 minutes,     Δt: 10 seconds,     wall time: 4.114 minutes,     adv CFL: 0.00e+00,     diff CFL: 0.00e+00
[ Info: [002.92%] i:    105,     time: 17.500 minutes,     Δt: 10 seconds,     wall time: 4.253 minutes,     adv CFL: 0.00e+00,     diff CFL: 0.00e+00
[ Info: [003.06%] i:    110,     time: 18.333 minutes,     Δt: 10 seconds,     wall time: 4.392 minutes,     adv CFL: 0.00e+00,     diff CFL: 0.00e+00
[ Info: [003.19%] i:    115,     time: 19.167 minutes,     Δt: 10 seconds,     wall time: 4.531 minutes,     adv CFL: 0.00e+00,     diff CFL: 0.00e+00
[ Info: [003.33%] i:    120,     time: 20.000 minutes,     Δt: 10 seconds,     wall time: 4.670 minutes,     adv CFL: 0.00e+00,     diff CFL: 0.00e+00
[ Info: [003.47%] i:    125,     time: 20.833 minutes,     Δt: 10 seconds,     wall time: 4.809 minutes,     adv CFL: 0.00e+00,     diff CFL: 0.00e+00
[ Info: [003.61%] i:    130,     time: 21.667 minutes,     Δt: 10 seconds,     wall time: 4.948 minutes,     adv CFL: 0.00e+00,     diff CFL: 0.00e+00
[ Info: [003.75%] i:    135,     time: 22.500 minutes,     Δt: 10 seconds,     wall time: 5.087 minutes,     adv CFL: 0.00e+00,     diff CFL: 0.00e+00
[ Info: [003.89%] i:    140,     time: 23.333 minutes,     Δt: 10 seconds,     wall time: 5.226 minutes,     adv CFL: 0.00e+00,     diff CFL: 0.00e+00
[ Info: [004.03%] i:    145,     time: 24.167 minutes,     Δt: 10 seconds,     wall time: 5.365 minutes,     adv CFL: 0.00e+00,     diff CFL: 0.00e+00
[ Info: [004.17%] i:    150,     time: 25.000 minutes,     Δt: 10 seconds,     wall time: 5.504 minutes,     adv CFL: 0.00e+00,     diff CFL: 0.00e+00
[ Info: [004.31%] i:    155,     time: 25.833 minutes,     Δt: 10 seconds,     wall time: 5.643 minutes,     adv CFL: 0.00e+00,     diff CFL: 0.00e+00
[ Info: [004.44%] i:    160,     time: 26.667 minutes,     Δt: 10 seconds,     wall time: 5.782 minutes,     adv CFL: 0.00e+00,     diff CFL: 0.00e+00
[ Info: [004.58%] i:    165,     time: 27.500 minutes,     Δt: 10 seconds,     wall time: 5.921 minutes,     adv CFL: 0.00e+00,     diff CFL: 0.00e+00
[ Info: [004.72%] i:    170,     time: 28.333 minutes,     Δt: 10 seconds,     wall time: 6.060 minutes,     adv CFL: 0.00e+00,     diff CFL: 0.00e+00
[ Info: [004.86%] i:    175,     time: 29.167 minutes,     Δt: 10 seconds,     wall time: 6.199 minutes,     adv CFL: 0.00e+00,     diff CFL: 0.00e+00
```

While for Julia 1.6 this is the output after the same amount of wall time:

```julia
[ Info: ---> Starting run!
[ Info: [000.14%] i:      5,     time: 50.000 seconds,     Δt: 10 seconds,     wall time: 3.453 minutes,     adv CFL: 0.00e+00,     diff CFL: 0.00e+00
[ Info: [000.28%] i:     10,     time: 1.667 minutes,     Δt: 10 seconds,     wall time: 4.388 minutes,     adv CFL: 0.00e+00,     diff CFL: 0.00e+00
[ Info: [000.42%] i:     15,     time: 2.500 minutes,     Δt: 10 seconds,     wall time: 5.269 minutes,     adv CFL: 0.00e+00,     diff CFL: 0.00e+00
[ Info: [000.56%] i:     20,     time: 3.333 minutes,     Δt: 10 seconds,     wall time: 6.150 minutes,     adv CFL: 0.00e+00,     diff CFL: 0.00e+00
```

Has someone else experienced this?

Any ideas as to what might be causing it? I really would like to upgrade my production ready scripts but there's no way I can do it until this issue is resolved unfortunately :/

**Comment by ali-ramadhan** [comment]
Thanks for reporting this! Definitely weird... I'll see if I can reproduce on a different machine.

**Comment by glwagner** [comment]
It might be useful to know the GPU and CUDA version being used (not sure).

**Comment by ali-ramadhan** [comment]
I ran the advection scheme benchmarks and comparing with some older Julia 1.5 results it definitely seems slower on the GPU. WENO5 used to only be ~3x slower than CenteredSecondOrder, but now it's 26x slower.

Not slow enough to be CUDA scalar operations so maybe the GPU compiler changed in some way that kernels calling/using WENO5 are compiling into suboptimal machine code? @maledt might have some ideas/suggestions but maybe we just have to profile and find the new bottleneck?

---

```
              Advection schemes relative performance (GPU)
┌───────────────┬────────────────────────┬──────────┬─────────┬─────────┐
│ Architectures │                Schemes │ slowdown │  memory │  allocs │
├───────────────┼────────────────────────┼──────────┼─────────┼─────────┤
│           GPU │    CenteredFourthOrder │  1.38356 │ 1.05911 │ 1.60067 │
│           GPU │    CenteredSecondOrder │      1.0 │     1.0 │     1.0 │
│           GPU │ UpwindBiasedFifthOrder │  1.53145 │  1.0868 │ 1.88203 │
│           GPU │ UpwindBiasedThirdOrder │  1.30611 │ 1.04135 │ 1.42012 │
│           GPU │                  WENO5 │  26.1429 │ 4.68526 │ 38.4468 │
└───────────────┴────────────────────────┴──────────┴─────────┴─────────┘
```

Compare with: https://github.com/CliMA/Oceananigans.jl/pull/1169#issuecomment-725471594

**Comment by glwagner** [comment]
Is the slow down different for different topologies? If so that might be a clue.

`CenteredSecondOrder` is special because it directly defines the function `advective_momentum_flux_Uu`:

https://github.com/CliMA/Oceananigans.jl/blob/383173d11a0c96182a4349fc1e33755207bf0886/src/Advection/centered_second_order.jl#L11

The other schemes define `symmetric_interpolate_*` and `left_biased_interpolate_*`, etc. For example, `CenteredFourthOrder`:

https://github.com/CliMA/Oceananigans.jl/blob/383173d11a0c96182a4349fc1e33755207bf0886/src/Advection/centered_fourth_order.jl#L21

These functions are filtered through an if statement if the dimension is `Bounded`, see:

https://github.com/CliMA/Oceananigans.jl/blob/master/src/Advection/topologically_conditional_interpolation.jl

**Comment by glwagner** [comment]
https://github.com/CliMA/Oceananigans.jl/pull/1733 changes the advection schemes a bit and adds the "boundary buffer" as an integer that's known at compile time.

Might be worth testing that PR since it changes the topological condition to use that type information. On master we have

https://github.com/CliMA/Oceananigans.jl/blob/383173d11a0c96182a4349fc1e33755207bf0886/src/Advection/topologically_conditional_interpolation.jl#L17-L22

whereas on https://github.com/CliMA/Oceananigans.jl/pull/1733 it's

https://github.com/CliMA/Oceananigans.jl/blob/17c01a9f3eca4e8576458f6c6f444f9cd2278cc3/src/Advection/topologically_conditional_interpolation.jl#L17-L22

**Comment by francispoulin** [comment]
Yes, I have seen the same thing, as discussed in #1722.

Previously, when looking at speed up on find grids, say 8000^2, I found that the speed up on GPUS was almost 400 times faster compared to the 5th order Upwinding.  When @hennyg888 did the same benchmark on julia 1.6, there was still speedup, but only by 200 or so.  The speed of the CPUS with 1.5 and 1.6 was very similar.  That means that WENO5, in this case, is about half as slow as it used to be.  

Sadly, the old data is lost in slack oblivion, but I suppose we could run old could and obtain these results again, if there was desire to do so.  But I think this discussion is more fruitful by looking at the nuts and bolts of the method.

Thanks @tomchor  for bringing this up as well!

**Comment by ali-ramadhan** [comment]
I ran the benchmark again with triply periodic but it's still much slower so the issue might be deeper than the logic in `topologically_conditional_interpolation.jl`.

```
              Advection schemes relative performance (GPU)
┌───────────────┬────────────────────────┬──────────┬─────────┬─────────┐
│ Architectures │                Schemes │ slowdown │  memory │  allocs │
├───────────────┼────────────────────────┼──────────┼─────────┼─────────┤
│           GPU │    CenteredFourthOrder │  1.50326 │ 1.06836 │ 1.69674 │
│           GPU │    CenteredSecondOrder │      1.0 │     1.0 │     1.0 │
│           GPU │ UpwindBiasedFifthOrder │  1.69787 │ 1.09472 │ 1.96539 │
│           GPU │ UpwindBiasedThirdOrder │  1.39899 │ 1.05598 │ 1.57057 │
│           GPU │                  WENO5 │  33.2728 │ 5.21273 │ 43.9286 │
└───────────────┴────────────────────────┴──────────┴─────────┴─────────┘
```

```diff
diff --git a/benchmark/benchmark_advection_schemes.jl b/benchmark/benchmark_advection_schemes.jl
index 81b083e1..e6ba8cd6 100644
--- a/benchmark/benchmark_advection_schemes.jl
+++ b/benchmark/benchmark_advection_schemes.jl
@@ -7,7 +7,8 @@ using Benchmarks
 # Benchmark function

 function benchmark_advection_scheme(Arch, Scheme)
-    grid = RegularRectilinearGrid(size=(192, 192, 192), extent=(1, 1, 1))
+    topo = (Periodic, Periodic, Periodic)
+    grid = RegularRectilinearGrid(topology=topo, size=(192, 192, 192), extent=(1, 1, 1))
     model = IncompressibleModel(architecture=Arch(), grid=grid, advection=Scheme())
 
     time_step!(model, 1) # warmup
```

**Comment by tomchor** [comment]
Yeah, I've run quite a bit of tests at this point, and the issue seems persistent and (as far as I could tell) independent of topology (although I haven't tried every single topology option).

Thanks for looking into this, btw. Let's hope it's something simple. Let me know how I can help.

**Comment by glwagner** [comment]
I guess another thing to check is if there is a similar slowdown with biharmonic diffusivity. In that case we might be able to pin the problem on the larger stencil, perhaps.

**Comment by francispoulin** [comment]
In #1722 we found  using 5th order Upwinding, on a grid with 128^3, that the speed up was about 80.  Is the above really saying its 1.69?  If so then things have certainly changed from 28 days ago.

**Comment by ali-ramadhan** [comment]
@francispoulin Ah 1.69 is how much slower `UpwindBiasedFifthOrder` is on the GPU instead of `CenteredSecondOrder` (also on the GPU). Below are the raw benchmarks and the CPU -> GPU speedups which show a speedup of ~114x for `UpwindBiasedFifthOrder` on 192^3 which should agree better with your figure of ~80x.

Actually looking at the advection scheme benchmarks more closely it looks like WENO5 is incurring lots of CPU allocations. According to https://github.com/CliMA/Oceananigans.jl/pull/1169#issuecomment-725471594 changing the advection did not change the number of allocations, but now it does and WENO5 allocates much more memory than the other schemes.

@glwagner I posted the turbulence closure benchmarks below and they seem fine/unchanged.

---

# Advection scheme benchmarks

```
                                                 Advection scheme benchmarks                                                                                                                                                                                
┌───────────────┬────────────────────────┬────────────┬────────────┬────────────┬────────────┬───────────┬────────┬─────────┐                                                                                                                               
│ Architectures │                Schemes │        min │     median │       mean │        max │    memory │ allocs │ samples │                                                                                                                              
├───────────────┼────────────────────────┼────────────┼────────────┼────────────┼────────────┼───────────┼────────┼─────────┤                                                                                                                              
│           CPU │    CenteredFourthOrder │    1.541 s │    1.545 s │    1.545 s │    1.548 s │  1.61 MiB │   2096 │       4 │                                                                                                                              
│           CPU │    CenteredSecondOrder │    1.029 s │    1.035 s │    1.036 s │    1.048 s │  1.61 MiB │   2096 │       5 │                                                                                                                              
│           CPU │ UpwindBiasedFifthOrder │    2.250 s │    2.251 s │    2.251 s │    2.252 s │  1.61 MiB │   2096 │       3 │                                                                                                                              
│           CPU │ UpwindBiasedThirdOrder │    1.589 s │    1.594 s │    1.594 s │    1.599 s │  1.61 MiB │   2096 │       4 │                                                                                                                              
│           CPU │                  WENO5 │    6.339 s │    6.339 s │    6.339 s │    6.339 s │  1.61 MiB │   2096 │       1 │                                                                                                                              
│           GPU │    CenteredFourthOrder │  17.309 ms │  17.419 ms │  18.107 ms │  24.384 ms │  2.71 MiB │  27650 │      10 │                                                                                                                              
│           GPU │    CenteredSecondOrder │  10.369 ms │  11.588 ms │  11.472 ms │  11.642 ms │  2.53 MiB │  16296 │      10 │
│           GPU │ UpwindBiasedFifthOrder │  19.561 ms │  19.675 ms │  20.975 ms │  32.694 ms │  2.77 MiB │  32028 │      10 │                 
│           GPU │ UpwindBiasedThirdOrder │  16.131 ms │  16.211 ms │  16.806 ms │  22.239 ms │  2.68 MiB │  25594 │      10 │
│           GPU │                  WENO5 │ 382.916 ms │ 385.558 ms │ 385.368 ms │ 386.709 ms │ 13.21 MiB │ 715860 │      10 │
└───────────────┴────────────────────────┴────────────┴────────────┴────────────┴────────────┴───────────┴────────┴─────────┘
```

```
          Advection schemes CPU to GPU speedup                                    
┌────────────────────────┬─────────┬─────────┬─────────┐          
│                Schemes │ speedup │  memory │  allocs │          
├────────────────────────┼─────────┼─────────┼─────────┤          
│    CenteredFourthOrder │ 88.7159 │  1.6849 │ 13.1918 │          
│    CenteredSecondOrder │ 89.3514 │ 1.57709 │ 7.77481 │          
│ UpwindBiasedFifthOrder │   114.4 │ 1.72647 │ 15.2805 │            
│ UpwindBiasedThirdOrder │ 98.3274 │ 1.66538 │ 12.2109 │                             
│                  WENO5 │ 16.4404 │ 8.22094 │ 341.536 │                         
└────────────────────────┴─────────┴─────────┴─────────┘
```

# Turbulence closure benchmarks

```
                  Turbulence closures relative performance (GPU)
┌───────────────┬──────────────────────────────────┬──────────┬─────────┬─────────┐
│ Architectures │                         Closures │ slowdown │  memory │  allocs │
├───────────────┼──────────────────────────────────┼──────────┼─────────┼─────────┤
│           GPU │ AnisotropicBiharmonicDiffusivity │   1.5313 │ 1.03189 │ 1.54697 │
│           GPU │           AnisotropicDiffusivity │  1.05623 │ 1.00582 │ 1.01779 │
│           GPU │    AnisotropicMinimumDissipation │  1.46265 │ 1.19908 │ 1.26817 │
│           GPU │             IsotropicDiffusivity │  1.13134 │ 1.00607 │ 1.07995 │
│           GPU │                          Nothing │      1.0 │     1.0 │     1.0 │
│           GPU │                 SmagorinskyLilly │  1.41905 │ 1.30373 │ 1.18683 │
│           GPU │              TwoDimensionalLeith │  1.11312 │ 1.06941 │ 1.06147 │
└───────────────┴──────────────────────────────────┴──────────┴─────────┴─────────┘
```

**Comment by francispoulin** [comment]
Thanks @ali-ramadhan for clarifying that makes much more sense.  A speed up for `WENO5` is evern worst than what we saw last month, which is pretty much the same as `UpwindBiasedFifthOrder`.  Hmm....

**Comment by glwagner** [comment]
I'm sorry, I misinterpreted the results @ali-ramadhan posted. I thought that `CenteredSecondOrder` was 1.0x slower with julia 1.6 than with 1.5 (and that small slowdowns were observed for the other schemes, which is why I recommended testing the biharmonic scheme.) Now I understand that these results are all for julia 1.6; we are comparing the results with previously obtained benchmarks (not posted) for julia 1.5.

Looking at @tomchor and @ali-ramadhan's results then it looks like simulations with WENO5 are running approximately 6-8 times slower on julia 1.6 than it was on julia 1.5, while other advection schemes (and closures) are unchanged --- correct?

Is the _CPU_ performance of WENO5 roughly equivalent between julia 1.5 and julia 1.6?

**Comment by tomchor** [comment]
Yes, from what I could test so far the CPU performance seems roughly
equivalent between versions. Although it would be good if someone else
tried to validate that as well.

**Comment by tomchor** [comment]
Has anyone tried or knows how to profile these functions? I feel that it would be much easier to find out what's wrong if we profile WENO5 in Julia 1.5 and 1.6 separately.

**Comment by glwagner** [comment]
Profiling is a very good idea. It probably makes sense to use an integrated / application profiler (rather than simply timing functions), because WENO5 is itself composed of many small functions and we don't know which one is the bottleneck.

I have never tried profiling on the GPU, but there's some info here: https://juliagpu.gitlab.io/CUDA.jl/development/profiling/

**Comment by glwagner** [comment]
Here's something: https://github.com/CliMA/Oceananigans.jl/pull/1770

I'm trying to run the benchmarks but they take a while, so that's in progress.

**Comment by glwagner** [comment]
I believe #1770 does the trick:

```
[2021/06/25 18:04:55.066] INFO  Writing Advection_schemes_relative_performance_(CPU).html...
              Advection schemes relative performance (GPU)
┌───────────────┬────────────────────────┬──────────┬─────────┬─────────┐
│ Architectures │                Schemes │ slowdown │  memory │  allocs │
├───────────────┼────────────────────────┼──────────┼─────────┼─────────┤
│           GPU │    CenteredFourthOrder │  1.36629 │ 1.07711 │ 1.66944 │
│           GPU │    CenteredSecondOrder │      1.0 │     1.0 │     1.0 │
│           GPU │ UpwindBiasedFifthOrder │  1.53522 │ 1.11266 │  1.9781 │
│           GPU │ UpwindBiasedThirdOrder │  1.31322 │ 1.03505 │ 1.30432 │
│           GPU │                  WENO5 │  1.84272 │  1.1889 │ 2.64008 │

```

would be good to get confirmation from someone.

**Comment by tomchor** [comment]
Confirmed and approved. Would be good to release a new version with this asap.

**Comment by hennyg888** [comment]
Just ran `benchmark_shallow_water_model.jl` with the fixes introduced in #1770. There is indeed a notable increase in speedup compared to no specified advection scheme as shown in #1722. The CPU to GPU speedup went up from ~180 times to ~400 times.

```
Oceananigans v0.58.1
Julia Version 1.6.0
Commit f9720dc2eb (2021-03-24 12:55 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: Intel(R) Xeon(R) Silver 4216 CPU @ 2.10GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-11.0.1 (ORCJIT, cascadelake)
Environment:
  EBVERSIONJULIA = 1.6.0
  JULIA_DEPOT_PATH = :
  EBROOTJULIA = /cvmfs/soft.computecanada.ca/easybuild/software/2020/avx2/Core/julia/1.6.0
  EBDEVELJULIA = /cvmfs/soft.computecanada.ca/easybuild/software/2020/avx2/Core/julia/1.6.0/easybuild/avx2-Core-julia-1.6.0-easybuild-devel
  JULIA_LOAD_PATH = :
  GPU: Tesla V100-SXM2-32GB

                                              Shallow water model benchmarks
┌───────────────┬─────────────┬───────┬────────────┬────────────┬────────────┬────────────┬───────────┬────────┬─────────┐
│ Architectures │ Float_types │    Ns │        min │     median │       mean │        max │    memory │ allocs │ samples │
├───────────────┼─────────────┼───────┼────────────┼────────────┼────────────┼────────────┼───────────┼────────┼─────────┤
│           CPU │     Float64 │    32 │   2.677 ms │   2.876 ms │   3.047 ms │   4.806 ms │  1.36 MiB │   2253 │      10 │
│           CPU │     Float64 │    64 │   5.795 ms │   5.890 ms │   6.073 ms │   7.770 ms │  1.36 MiB │   2255 │      10 │
│           CPU │     Float64 │   128 │  16.979 ms │  17.350 ms │  17.578 ms │  19.993 ms │  1.36 MiB │   2255 │      10 │
│           CPU │     Float64 │   256 │  62.543 ms │  63.222 ms │  63.544 ms │  67.347 ms │  1.36 MiB │   2255 │      10 │
│           CPU │     Float64 │   512 │ 250.149 ms │ 251.023 ms │ 251.092 ms │ 252.389 ms │  1.36 MiB │   2315 │      10 │
│           CPU │     Float64 │  1024 │ 990.901 ms │ 993.115 ms │ 993.360 ms │ 996.091 ms │  1.36 MiB │   2315 │       6 │
│           CPU │     Float64 │  2048 │    4.002 s │    4.004 s │    4.004 s │    4.007 s │  1.36 MiB │   2315 │       2 │
│           CPU │     Float64 │  4096 │   16.371 s │   16.371 s │   16.371 s │   16.371 s │  1.36 MiB │   2315 │       1 │
│           CPU │     Float64 │  8192 │   64.657 s │   64.657 s │   64.657 s │   64.657 s │  1.36 MiB │   2315 │       1 │
│           CPU │     Float64 │ 16384 │  290.423 s │  290.423 s │  290.423 s │  290.423 s │  1.36 MiB │   2315 │       1 │
│           GPU │     Float64 │    32 │   3.468 ms │   3.656 ms │   3.745 ms │   4.695 ms │  1.82 MiB │   5687 │      10 │
│           GPU │     Float64 │    64 │   3.722 ms │   3.903 ms │   4.050 ms │   5.671 ms │  1.82 MiB │   5687 │      10 │
│           GPU │     Float64 │   128 │   3.519 ms │   3.808 ms │   4.042 ms │   6.372 ms │  1.82 MiB │   5687 │      10 │
│           GPU │     Float64 │   256 │   3.822 ms │   4.153 ms │   4.288 ms │   5.810 ms │  1.82 MiB │   5687 │      10 │
│           GPU │     Float64 │   512 │   4.637 ms │   4.932 ms │   4.961 ms │   5.728 ms │  1.82 MiB │   5765 │      10 │
│           GPU │     Float64 │  1024 │   3.240 ms │   3.424 ms │   3.527 ms │   4.553 ms │  1.82 MiB │   5799 │      10 │
│           GPU │     Float64 │  2048 │  10.783 ms │  10.800 ms │  11.498 ms │  17.824 ms │  1.98 MiB │  16305 │      10 │
│           GPU │     Float64 │  4096 │  41.880 ms │  41.911 ms │  42.485 ms │  47.627 ms │  2.67 MiB │  61033 │      10 │
│           GPU │     Float64 │  8192 │ 166.751 ms │ 166.800 ms │ 166.847 ms │ 167.129 ms │  5.21 MiB │ 227593 │      10 │
│           GPU │     Float64 │ 16384 │ 681.129 ms │ 681.249 ms │ 681.301 ms │ 681.583 ms │ 16.59 MiB │ 973627 │       8 │
└───────────────┴─────────────┴───────┴────────────┴────────────┴────────────┴────────────┴───────────┴────────┴─────────┘

        Shallow water model CPU to GPU speedup
┌─────────────┬───────┬──────────┬─────────┬─────────┐
│ Float_types │    Ns │  speedup │  memory │  allocs │
├─────────────┼───────┼──────────┼─────────┼─────────┤
│     Float64 │    32 │ 0.786715 │ 1.33777 │ 2.52419 │
│     Float64 │    64 │  1.50931 │ 1.33774 │ 2.52195 │
│     Float64 │   128 │  4.55587 │ 1.33774 │ 2.52195 │
│     Float64 │   256 │  15.2238 │ 1.33774 │ 2.52195 │
│     Float64 │   512 │  50.8995 │ 1.33771 │ 2.49028 │
│     Float64 │  1024 │  290.085 │ 1.33809 │ 2.50497 │
│     Float64 │  2048 │  370.777 │ 1.45575 │  7.0432 │
│     Float64 │  4096 │  390.617 │ 1.95667 │ 26.3641 │
│     Float64 │  8192 │  387.632 │ 3.82201 │ 98.3123 │
│     Float64 │ 16384 │   426.31 │  12.177 │ 420.573 │
└─────────────┴───────┴──────────┴─────────┴─────────┘
```

**Comment by glwagner** [comment]
thanks @hennyg888 .
